// Decompiled with JetBrains decompiler
// Type: SmartAssembly.SmartExceptionsCore.UnhandledException
// Assembly: fe64d8c2-dc44-4388-ab3c-b2f70f696e6b, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: AAC993D7-E494-4D2E-8994-D4537A031819
// Assembly location: C:\Users\Administrateur\Downloads\VirusShare_CryptoRansom_20160715-msil\HEUR-Trojan.Win32.Generic-57b3b74addf88b0a5a3696c8126023d8603e667524dc220b61597349503669b3.exe

using \u0005;
using \u0007;
using \u0080;
using System;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;

namespace SmartAssembly.SmartExceptionsCore
{
  [Serializable]
  public class UnhandledException : Exception
  {
    [NonSerialized]
    internal static \u0004 \u0001;
    public int MethodID;
    public object[] Objects;
    public int ILOffset;
    private Exception previousException;

    internal Exception \u0013() => this.previousException;

    public static Exception \u008D\u0002([In] Exception obj0) => UnhandledException.\u0098\u0002(obj0, new object[0]);

    public static Exception \u008E\u0002([In] Exception obj0, [In] object obj1) => UnhandledException.\u0098\u0002(obj0, new object[1]
    {
      obj1
    });

    public static Exception \u008F\u0002([In] Exception obj0, [In] object obj1, [In] object obj2) => UnhandledException.\u0098\u0002(obj0, new object[2]
    {
      obj1,
      obj2
    });

    public static Exception \u0090\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[3]
      {
        obj1,
        obj2,
        obj3
      });
    }

    public static Exception \u0091\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[4]
      {
        obj1,
        obj2,
        obj3,
        obj4
      });
    }

    public static Exception \u0092\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[5]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5
      });
    }

    public static Exception \u0093\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5,
      [In] object obj6)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[6]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5,
        obj6
      });
    }

    public static Exception \u0094\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5,
      [In] object obj6,
      [In] object obj7)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[7]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5,
        obj6,
        obj7
      });
    }

    public static Exception \u0095\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5,
      [In] object obj6,
      [In] object obj7,
      [In] object obj8)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[8]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5,
        obj6,
        obj7,
        obj8
      });
    }

    public static Exception \u0096\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5,
      [In] object obj6,
      [In] object obj7,
      [In] object obj8,
      [In] object obj9)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[9]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5,
        obj6,
        obj7,
        obj8,
        obj9
      });
    }

    public static Exception \u0097\u0002(
      [In] Exception obj0,
      [In] object obj1,
      [In] object obj2,
      [In] object obj3,
      [In] object obj4,
      [In] object obj5,
      [In] object obj6,
      [In] object obj7,
      [In] object obj8,
      [In] object obj9,
      [In] object obj10)
    {
      return UnhandledException.\u0098\u0002(obj0, new object[10]
      {
        obj1,
        obj2,
        obj3,
        obj4,
        obj5,
        obj6,
        obj7,
        obj8,
        obj9,
        obj10
      });
    }

    public static Exception \u0098\u0002([In] Exception obj0, [In] object[] obj1)
    {
      if (\u0084.\u0002)
        return (Exception) null;
      int num1 = -1;
      int num2 = -1;
      try
      {
        StackTrace stackTrace = new StackTrace(obj0);
        if (\u0014\u0005.\u007E\u0087\u0004((object) stackTrace) > 0)
        {
          StackFrame stackFrame = \u008D\u0004.\u007E\u0088\u0004((object) stackTrace, \u0014\u0005.\u007E\u0087\u0004((object) stackTrace) - 1);
          num2 = (\u0014\u0005.\u007E\u0083\u0003((object) \u0017\u0003.\u007E\u0089\u0004((object) stackFrame)) & 16777215) - 1;
          num1 = \u0014\u0005.\u007E\u008A\u0004((object) stackFrame);
        }
      }
      catch
      {
      }
      UnhandledException unhandledException = new UnhandledException(num2, obj1, num1, obj0);
      if (obj0 is UnhandledException)
      {
        Exception exception = (obj0 as UnhandledException).\u0013();
        if (exception != null)
          obj0 = exception;
      }
      Exception exception1 = obj0;
      while (\u0097\u0003.\u007E\u001F\u0002((object) exception1) != null)
        exception1 = \u0097\u0003.\u007E\u001F\u0002((object) exception1);
      try
      {
        // ISSUE: type reference
        FieldInfo fieldInfo = \u0004\u0005.\u007E\u008F\u0003((object) \u008B\u0005.\u0088\u0003(__typeref (Exception)), UnhandledException.\u0001(3519), BindingFlags.Instance | BindingFlags.NonPublic);
        \u0093\u0004.\u007E\u0007\u0005((object) fieldInfo, (object) exception1, (object) unhandledException);
      }
      catch
      {
      }
      return obj0;
    }

    public override void GetObjectData([In] SerializationInfo obj0, [In] StreamingContext obj1)
    {
      \u0013\u0004.\u0080\u0002((object) this, obj0, obj1);
      // ISSUE: type reference
      \u008E\u0003.\u007E\u0013\u0005((object) obj0, UnhandledException.\u0001(3540), (object) this.MethodID, \u008B\u0005.\u0088\u0003(__typeref (int)));
      // ISSUE: type reference
      \u008E\u0003.\u007E\u0013\u0005((object) obj0, UnhandledException.\u0001(3577), (object) this.ILOffset, \u008B\u0005.\u0088\u0003(__typeref (int)));
      // ISSUE: type reference
      \u008E\u0003.\u007E\u0013\u0005((object) obj0, UnhandledException.\u0001(3614), (object) this.previousException, \u008B\u0005.\u0088\u0003(__typeref (Exception)));
      int num = this.Objects == null ? 0 : this.Objects.Length;
      // ISSUE: type reference
      \u008E\u0003.\u007E\u0013\u0005((object) obj0, UnhandledException.\u0001(3663), (object) num, \u008B\u0005.\u0088\u0003(__typeref (int)));
      for (int index = 0; index < num; ++index)
      {
        // ISSUE: type reference
        \u008E\u0003.\u007E\u0013\u0005((object) obj0, \u0011\u0005.\u000F\u0002(UnhandledException.\u0001(3708), (object) index), this.Objects[index], \u008B\u0005.\u0088\u0003(__typeref (object)));
      }
    }

    internal UnhandledException([In] SerializationInfo obj0, [In] StreamingContext obj1)
      : base(obj0, obj1)
    {
      this.MethodID = \u0006\u0003.\u007E\u0015\u0005((object) obj0, UnhandledException.\u0001(3540));
      this.ILOffset = \u0006\u0003.\u007E\u0015\u0005((object) obj0, UnhandledException.\u0001(3577));
      // ISSUE: type reference
      this.previousException = (Exception) \u0003\u0004.\u007E\u0014\u0005((object) obj0, UnhandledException.\u0001(3614), \u008B\u0005.\u0088\u0003(__typeref (Exception)));
      int length = \u0006\u0003.\u007E\u0015\u0005((object) obj0, UnhandledException.\u0001(3663));
      this.Objects = new object[length];
      for (int index = 0; index < length; ++index)
      {
        // ISSUE: type reference
        this.Objects[index] = \u0003\u0004.\u007E\u0014\u0005((object) obj0, \u0011\u0005.\u000F\u0002(UnhandledException.\u0001(3708), (object) index), \u008B\u0005.\u0088\u0003(__typeref (object)));
      }
    }

    internal UnhandledException([In] int obj0, [In] object[] obj1, [In] int obj2, [In] Exception obj3)
      : base(string.Format(UnhandledException.\u0001(3753), (object) obj0, (object) obj2))
    {
      this.MethodID = obj0;
      this.Objects = obj1;
      this.ILOffset = obj2;
      this.previousException = obj3;
    }

    static UnhandledException() => \u0006.\u0003\u0002();
  }
}
